### PR TITLE
parallel csv scan

### LINKED ISSUE #6325: Parallel CSV reading
### Is your feature request related to a problem or challenge?

As part of having a great "out of the box" experience it is important to use as many cores as possible in DataFusion. Given modern consumer laptops have 8-16 cores using multiple cores can literally translate to an order of magnitude faster performance.

While DataFusion offers the ability to read partitioned datasets (aka when the input is in multiple files), often, especially for initially testing out the tool,  people will simply running queries on their existing CSV or JSON datasets and it will be relatively slow.

We already have the great `datafusion.optimizer.repartition_file_scans` option (see [docs](https://arrow.apache.org/datafusion/user-guide/configs.html))  -- added by @korowa  in https://github.com/apache/arrow-datafusion/pull/5057 (👋 !) which uses multiple cores to decode the parquet files in parallel. I would like a similar feature for CSV files

### Describe the solution you'd like

One basic approach (following what @korowa  did for Parquet) would be:
1. If the `datafusion.optimizer.repartition_file_scans` option is set, divide the file into even (byte) sized contiguous blocks, probably with some lower limit (like 1MB)
3. Update [CsvExec](https://github.com/apache/arrow-datafusion/blob/8a25953182b361a58c56384129bf57f32aa2dbb1/datafusion/core/src/physical_plan/file_format/csv.rs#L53)  to process partitions using those subsets of the viles

Notes:
Given the vagaries of CSV (e.g. unescaped quoted newlines) it is likely impossible to parallelize CSV reading for all possible files. I think this is fine, and as long as we can turn off the reading in parallel it is better to have faster out of the box query performance for 99.99% of the queries than handle bizzare CSV files always

Care will be required to make sure all records are read exactly once, given the partition splits will likely be in the middle of rows. 

One idea for parsing a partition (`offset`,  `len`):
1. Start CSV parsing the data starting *after* finding the next newline after the `offset` bytes
2. Continue CSV parsing until the newline *after* the `offset + len` byte 

```
        0        A,1,2,3,4,5,6,7,8,9\n                            
        20       A,1,2,3,4,5,6,7,8,9\n                            
        40       A,1,2,3,4,5,6,7,8,9\n ◀─ ─ ─ ─ ─ ─ ─ ─           
        60       A,1,2,3,4,5,6,7,8,9\n                 │          
        80       A,1,2,3,4,5,6,7,8,9\n                            
        100      A,1,2,3,4,5,6,7,8,9\n                 │          
                                                                  
Byte Offset       Lines of CSV Data                    │          
                  (in this case 20                                
                  bytes per line)           Split at byte 50 is in
                                              the middle of this  
                                                     line         
                                                                  
```
This is similar to what is described in https://github.com/apache/arrow-datafusion/issues/5205#issuecomment-1420567792

### Describe alternatives you've considered

_No response_

### Additional context

@kmitchener  noticed the same thing in: https://github.com/apache/arrow-datafusion/issues/5205

The duckdb implementation of a similar feature may offer some inspiration: https://github.com/duckdb/duckdb/pull/5194

### PR DESCRIPTION
**Pull request by 2010YOUY01**
# Which issue does this PR close?

Closes #6325.

# Rationale for this change

There are 3 steps to scan a CSV file in parallel:
* The initial physical plan contains some whole file scans during the planing
  * `CsvExec: file_groups={1 group: [[a.csv:0..100]]} `
* **Step 1:** `PhysicalOptimizerRule` (`Repartition`) will decide if the `CsvExec` node can be parallelized, it won't be parallelized if the parent node requires certain order or have some other conditions.
* **Step 2:** Repartition the byte range evenly during physical optimization `Repartition`, and store partitions in `CsvExec.base_config.file_groups`, also this step doesn't care about how to separate lines correctly (partitions may contain half lines)
  * `CsvExec: file_groups={2 groups: [[a.csv:0..50], [a:50..100]]}`
* **Step 3:** A worker will be spawned for each file group partition (e.g. `[a.csv:0..50]`), and deal with partition boundaries
  * (`[a.csv:0..50] -> Lines:1..5`).

# What changes are included in this PR?
### Step 1
The parallel parquet scan PR has already done it https://github.com/apache/arrow-datafusion/pull/5057, for parallel CSV scan, it only added a rule to not repartition if the CSV file is compressed
Testing - Added `CsvExec` case along with the existing tests for optimizing rule `Repartition` on `ParquetExec`
### Step 2
Parallel Parquet PR also has done this but inside `ParquetExec`, now the repartition byte range logic is refactored to somewhere else to let `CsvExec` reuse it.
Testing - Unit test for partitioning byte range has also been well covered by the parallel Parquet PR https://github.com/apache/arrow-datafusion/blob/e91af991c5ae4a6b4afab2cb1b0c9307a69e4046/datafusion/core/src/datasource/physical_plan/parquet.rs#L1924-L2128
### Step 3
The logic for handling line boundaries correctly is done in `CsvOpener`, the rule used is exactly the same as https://github.com/apache/arrow-datafusion/issues/6325 mentioned. Unlike parallel Parquet scan, which can use metadata to decide which row groups to read according to the approximate byte range, CSV scan needs to inspect the file content around partition boundaries to determine the byte range of lines in this partition. In the implementation, it finds the offset to the first newline encountered from the approximate byte partition start or end, calculates the byte range for complete lines of this partition, and reads that actual byte range from the object store.
Testing - Added some integration tests in `file_format/csv.rs` (also manually run these tests on S3). For more complex query tests, the existing TPCH correctness test under sqllogictest is scanned on CSV files with 4 partitions.
Some issue:
1. Range get not working for local filesystem https://github.com/apache/arrow-rs/blob/0d4e6a727f113f42d58650d2dbecab89b22d4e28/object_store/src/lib.rs#L355, need to update implementation after it's fixed
2. Need 3 get requests to object store for each partition (find head offset, find tail offset, and finally actually get the correct byte range for lines) It's easier to do this way but 3 * `target_partitions` requests on cloud stores(billed by number of requests) might be a problem. This may be improved by some follow on PR, but I haven't come up with a good solution yet, arrow CSV reader now only can handle valid CSV files. Maybe add a wrapper on the byte stream fetched from the object store, and let it skip the first line and read until the first newline after the range length?

# Are these changes tested?
See above section.
A very simple benchmark: scan TPC-H lineitem table (SF1) with very selective predicate (`select * from lineitem where column_1 = 1;`). 6 is # of physical cores for my machine
```
1 partitions: 54.127 s
2 partitions: 27.176 s
3 partitions: 20.532 s
4 partitions: 16.147 s
5 partitions: 15.061 s
6 partitions: 13.396 s
```

# Are there any user-facing changes?
No

### CODE REVIEW

**Comment by alamb** [comment]
Thank you @2010YOUY01  -- I will review this over the coming days. This looks awesoem

File: datafusion/core/Cargo.toml [hunk 1]
```diff
@@ -84,6 +84,7 @@ parquet = { workspace = true }
 percent-encoding = "2.2.0"
 pin-project-lite = "^0.2.7"
 rand = "0.8"
+regex = "1.5.4"
```
Comment (alamb) [inline datafusion/core/Cargo.toml:87]: since this is only used for tests, I think it can be put in the `dev-dependencies` section

**Comment by alamb** [comment]
I started running some tests with this branch. I will continue tomorrow

**Comment by alamb** [comment]
This is on my list to review, but I might not have time for it until later in the week (Wednesday) -- I am somewhat occupied with #6800  at the moment

**Comment by alamb** [comment]
Hi @2010YOUY01  -- I am having trouble reproducing the benchmark results you reported

# Results
Master:
```
❯ select count(*) from lineitem where l_quantity < 10;
1 row in set. Query took 1.424 seconds.
1 row in set. Query took 1.374 seconds.
1 row in set. Query took 1.409 seconds.
```

This PR branch:
```
❯ select count(*) from lineitem where l_quantity < 10;
1 row in set. Query took 1.918 seconds.
1 row in set. Query took 1.672 seconds.
1 row in set. Query took 2.008 seconds.
```
(I also merged up your branch from master and it still had the same performance)

# Methodology:

I tested this branch out using the TPCH SF1 (6M rows,  725 MB) lineitem CSV file (created with `arrow-datafusion/benchmarks$ ./bench.sh data tpch`):

```shell
(arrow_dev) alamb@MacBook-Pro-8:~$ du -h  /Users/alamb/Software/arrow-datafusion/benchmarks/data/lineitem.tbl
725M	/Users/alamb/Software/arrow-datafusion/benchmarks/data/lineitem.tbl

(arrow_dev) alamb@MacBook-Pro-8:~$ wc -l /Users/alamb/Software/arrow-datafusion/benchmarks/data/lineitem.tbl
 6001215 /Users/alamb/Software/arrow-datafusion/benchmarks/data/lineitem.tbl
```

And used `datafusion-cli` (built via `cargo build --release`)

```sql
CREATE EXTERNAL TABLE lineitem (
        l_orderkey BIGINT,
        l_partkey BIGINT,
        l_suppkey BIGINT,
        l_linenumber INTEGER,
        l_quantity DECIMAL(15, 2),
        l_extendedprice DECIMAL(15, 2),
        l_discount DECIMAL(15, 2),
        l_tax DECIMAL(15, 2),
        l_returnflag VARCHAR,
        l_linestatus VARCHAR,
        l_shipdate DATE,
        l_commitdate DATE,
        l_receiptdate DATE,
        l_shipinstruct VARCHAR,
        l_shipmode VARCHAR,
        l_comment VARCHAR,
        l_rev VARCHAR,
) STORED AS CSV DELIMITER '|' LOCATION '/Users/alamb/Software/arrow-datafusion/benchmarks/data/lineitem.tbl';

--- Run a query that scans the entire CSV
select count(*) from lineitem where l_quantity < 10;

+-----------------+
| COUNT(UInt8(1)) |
+-----------------+
| 1079240         |
+-----------------+
```

File: datafusion/core/src/datasource/physical_plan/mod.rs [hunk 2]
```diff
@@ -240,6 +242,81 @@ impl FileScanConfig {
...
+        }
+
+        let target_partition_size =
+            (total_size as usize + (target_partitions) - 1) / (target_partitions);
+
+        let current_partition_index: usize = 0;
+        let current_partition_size: usize = 0;
+
+        // Partition byte range evenly for all `PartitionedFile`s
+        let repartitioned_files = flattened_files
```
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/mod.rs:281]: One thing I was thinking about for the range based approach is that it isn't likely to work for streaming compressed files (as you need to decompress the data linearly)

I wonder if you have considered that 🤔

**Comment by 2010YOUY01** [comment]
> Hi @2010YOUY01 -- I am having trouble reproducing the benchmark results you reported

@alamb Thank you for the feedback!
My initial benchmark run the query under different `target_partition`, just realized that was not effective 😰 
I reproduced your benchmark. Since streaming byte range get on local FS is not implemented in Arrow yet, 
>Some issue:
>1. Range get not working for local filesystem https://github.com/apache/arrow-rs/blob/0d4e6a727f113f42d58650d2dbecab89b22d4e28/object_store/src/lib.rs#L355, need to update implementation after it's fixed

alternative `get_range()` is used (which will copy the range into memory at once instead of in a streaming fashion). It is called when finding the first newline after the start/end of each partition, and multiple unnecessary large disk read caused the performance issue.
This should be solved after `get_opts()` - `Range` option is supported for local FS. For now, I suppressed this issue with a preset max line length, and re-run the benchmark again:
This PR:
```
❯ select count(*) from lineitem where l_quantity < 10;
1 row in set. Query took 0.894 seconds.
1 row in set. Query took 0.513 seconds.
1 row in set. Query took 0.532 seconds.
```
Main branch:
```
❯ select count(*) from lineitem where l_quantity < 10;
1 row in set. Query took 1.757 seconds.
1 row in set. Query took 1.496 seconds.
1 row in set. Query took 1.498 seconds.
```

File: datafusion/core/src/datasource/physical_plan/mod.rs (hunk 2 above)
Comment (2010YOUY01) [reply datafusion/core/src/datasource/physical_plan/mod.rs:281]: Yes. For now, the physical optimizer will check that, and won't get compressed CSV files repartitioned.

**Comment by alamb** [comment]
Thanks @2010YOUY01  -- I am starting to look at this again

File: datafusion/core/src/datasource/physical_plan/parquet.rs [hunk 3]
```diff
@@ -253,75 +244,23 @@ impl ParquetExec {
...
-            .iter()
-            .flatten()
-            .collect::<Vec<_>>();
-
-        // Perform redistribution only in case all files should be read from beginning to end
-        let has_ranges = flattened_files.iter().any(|f| f.range.is_some());
-        if has_ranges {
-            return self.clone();
-        }
+        let repartitioned_file_groups_option = FileScanConfig::repartition_file_groups(
```
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/parquet.rs:253]: 👍 for this refactor

File: datafusion/core/src/datasource/physical_plan/parquet.rs [hunk 4]
```diff
@@ -1921,6 +1861,103 @@ mod tests {
...
+            .base_config()
+            .file_groups
+            .clone();
+
+        assert!(partitioned_file[0][0].range.is_none());
+    }
+
+    // Repartition when there is a empty file in file groups
+    #[tokio::test]
+    async fn parquet_exec_repartition_empty_files() {
```
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/parquet.rs:1897]: maybe as a follow on the tests for file repartitioning could be moved to datafusion/core/src/datasource/physical_plan/mod.rs to be in the same module they are testing (as they are now no longer specific to parquet)

File: datafusion/core/tests/sqllogictests/test_files/tpch/q1.slt.part [hunk 5]
```diff
@@ -57,9 +57,7 @@ SortPreservingMergeExec: [l_returnflag@0 ASC NULLS LAST,l_linestatus@1 ASC NULLS
...
 ----------------CoalesceBatchesExec: target_batch_size=8192
 ------------------FilterExec: l_shipdate@6 <= 10471
---------------------RepartitionExec: partitioning=RoundRobinBatch(4), input_partitions=1
-----------------------CsvExec: file_groups={1 group: [[WORKSPACE_ROOT/datafusion/core/tests/sqllogictests/test_files/tpch/data/lineitem.tbl]]}, projection=[l_quantity, l_extendedprice, l_discount, l_tax, l_returnflag, l_linestatus, l_shipdate], has_header=false
-
+--------------------CsvExec: file_groups={4 groups: [[WORKSPACE_ROOT/datafusion/core/tests/sqllogictests/test_files/tpch/data/lineitem.tbl:0..18561749], [WORKSPACE_ROOT/datafusion/core/tests/sqllogictests/test_files/tpch/data/lineitem.tbl:18561749..37123498], [WORKSPACE_ROOT/datafusion/core/tests/sqllogictests/test_files/tpch/data/lineitem.tbl:37123498..55685247], [WORKSPACE_ROOT/datafusion/core/tests/sqllogictests/test_files/tpch/data/lineitem.tbl:55685247..74246996]]}, projection=[l_quantity, l_extendedprice, l_discount, l_tax, l_returnflag, l_linestatus, l_shipdate], has_header=false
```
Comment (alamb) [inline datafusion/core/tests/sqllogictests/test_files/tpch/q1.slt.part:60]: ❤️

File: datafusion/core/src/physical_optimizer/repartition.rs [hunk 6]
```diff
@@ -257,6 +257,17 @@ fn optimize_partitions(
...
     }
 
+    if let Some(csv_exec) = new_plan.as_any().downcast_ref::<CsvExec>() {
+        // The underlying CsvOpener will only fetch certain part of csv file from the object store, which can't be decompressed separately
+        if repartition_file_scans && !csv_exec.file_compression_type.is_compressed() {
+            let repartitioned_exec_option =
+                csv_exec.get_repartitioned(target_partitions, repartition_file_min_size);
+            if let Some(repartitioned_exec) = repartitioned_exec_option {
+                return Ok(Transformed::Yes(Arc::new(repartitioned_exec)));
+            }
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/repartition.rs:267]: I recommend moving the check for compression `csv_exec.file_compression_type.is_compressed()` into `CsvExec::get_repartitioned()` to keep the logic for handling CSV files together. I don't think this is required however.

```suggestion
        if repartition_file_scans  {
            let repartitioned_exec_option =
                csv_exec.get_repartitioned(target_partitions, repartition_file_min_size);
            if let Some(repartitioned_exec) = repartitioned_exec_option {
                return Ok(Transformed::Yes(Arc::new(repartitioned_exec)));
            }
```

File: datafusion/core/src/datasource/physical_plan/csv.rs [hunk 7]
```diff
@@ -270,14 +297,223 @@ impl CsvOpener {
...
+}
+
 impl FileOpener for CsvOpener {
+    /// Open a partitioned CSV file.
+    ///
+    /// If `file_meta.range` is `None`, the entire file is opened.
+    /// If `file_meta.range` is `Some(FileRange {start, end})`, this signifies that the partition
+    /// corresponds to the byte range [start, end) within the file.
+    ///
+    /// Note: `start` or `end` might be in the middle of some lines. In such cases, the following rules
```
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/csv.rs:396]: As I understand it, this code does potentially several object store requests to adjust the initial ranges based on where the end of the CSV lines actually fall:

# Initial situation
```
CSV data with the next newlines (\n) after range marked                             
                                                                                    
 ┌─────────────┬──┬────────────────────┬──┬────────────────┬──┬────────┬──┬────────┐
 │     ...     │\n│        ...         │\n│      ...       │\n│  ...   │\n│  ...   │
 └─────────────┴──┴────────────────────┴──┴────────────────┴──┴────────┴──┴────────┘
        ▲                    ▲                     ▲               ▲                
        └─────────────┬──────┴─────────────────────┴───────────────┘                
                      │                                                             
                      │                                                             
             Initial file_meta.range es                                             
```

# This PR
```
    This PR: adjust the ranges prior to IO start, via                                
  object store operations                                                          
                                                                                   
┌ ─ ─ ─ ─ ─ ─ ─ ─┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┬ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┬ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ 
┌─────────────┬──┼────────────────────┬──┬────────────────┬──┬────────┬──┬────────┤
│     ...     │\n│        ...         │\n│      ...       │\n│  ...   │\n│  ...   │
└─────────────┴──┼────────────────────┴──┴────────────────┴──┴────────┴──┴────────┤
│                │                       │                   │                     
 ─ ─ ─ ─ ─ ─ ─ ─ ┘─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘
                                                                                   
 Partition 0       Partition 1            Partition 2       Partition 3            
 Read              Read                   Read              Read                   
```

This design has the nice property that each partition reads exactly the bytes it needs

This design has the downside that it requires several object store reads to find the newlines, and the overhead of each object store operation often is much larger than the overhead of reading extra data. 

# Alternate idea

Another approach that would reduce the number of object store requests would be to read past the initial range and stop at the next newline `\n` like this:

```
  Each partition reads *more* than its assigned ranged to find                     
  the trailing new line, and ignores everything afterwards                         
                                                                                   
                                                                                   
┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─                                                          
┌─────────────┬──┬───────┼────────────┬──┬────────────────┬──┬────────┬──┬────────┐
│     ...     │\n│        ...         │\n│      ...       │\n│  ...   │\n│  ...   │
└─────────────┴──┴───────┼────────────┴──┴────────────────┴──┴────────┴──┴────────┘
│                                                                                  
 ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘                                                         
  Partition 0                                                                      
  Read                                                                             
```

I think the tricky bit of this design would be to ensure enough extra data was read. Initially, maybe we could just pick something sufficiently large for most files, like 1MB and error if the next newline can't be found. As a follow on we could add some fanciness like make another object store request if necessary.

**Review by alamb — APPROVED** [review]
Thank you @2010YOUY01  -- I tried this out again and it does indeed go (much!) faster -- 3x faster in my initial testing. 👏 

I also reviewed the code, and I found it very easy to read,  well structured and well tested. Thank you so much for this and I am sorry for the delay. 

I have some thoughts on how to improve the byte range calculations to avoid the number of object store requests. I think this could be done as a follow on PR as well. 

My proposal is to:
1. Review the comments on this PR and consider if you want to make any more changes 
1. Merge this PR
2. File tickets for any remaining items
3. File tickets for parallel reading of compressed CSV / JSON files (I will do this)
4. File ticket for parallel reading of JSON files


# Testing Results


This branch now goes about 3x faster (3.9s vs 13.4s) on my test:

```shell
$ datafusion-cli -f /tmp/test.sql
DataFusion CLI v27.0.0
0 rows in set. Query took 0.001 seconds.
+-----------------+
| COUNT(UInt8(1)) |
+-----------------+
| 10796819        |
+-----------------+
1 row in set. Query took 3.961 seconds.
```

main:

```shell
$ ~/Software/target-df/release/datafusion-cli -f /tmp/test.sql
DataFusion CLI v27.0.0
0 rows in set. Query took 0.001 seconds.
+-----------------+
| COUNT(UInt8(1)) |
+-----------------+
| 10796819        |
+-----------------+
1 row in set. Query took 13.364 seconds.
```

Here is the queryL:

```sql
$ cat /tmp/test.sql
CREATE EXTERNAL TABLE lineitem (
        l_orderkey BIGINT,
        l_partkey BIGINT,
        l_suppkey BIGINT,
        l_linenumber INTEGER,
        l_quantity DECIMAL(15, 2),
        l_extendedprice DECIMAL(15, 2),
        l_discount DECIMAL(15, 2),
        l_tax DECIMAL(15, 2),
        l_returnflag VARCHAR,
        l_linestatus VARCHAR,
        l_shipdate DATE,
        l_commitdate DATE,
        l_receiptdate DATE,
        l_shipinstruct VARCHAR,
        l_shipmode VARCHAR,
        l_comment VARCHAR,
        l_rev VARCHAR,
) STORED AS CSV DELIMITER '|' LOCATION '/Users/alamb/Software/arrow-datafusion/benchmarks/data/tpch_sf10/lineitem.tbl';

--- Run a query that scans the entire CSV
select count(*) from lineitem where l_quantity < 10;
```

**Comment by 2010YOUY01** [comment]
> Thank you @2010YOUY01 -- I tried this out again and it does indeed go (much!) faster -- 3x faster in my initial testing. 👏

Thank you for the review ❤️  I will update this PR as per the comments next week

File: datafusion/core/src/datasource/physical_plan/csv.rs (hunk 7 above)
Comment (2010YOUY01) [reply datafusion/core/src/datasource/physical_plan/csv.rs:396]: These text charts look awesome 😮 How did you draw that?

**Comment by 2010YOUY01** [comment]
memo for myself: update comments/docs in configurations

File: datafusion/core/src/datasource/physical_plan/csv.rs (hunk 7 above)
Comment (alamb) [reply datafusion/core/src/datasource/physical_plan/csv.rs:396]: I use https://monodraw.helftone.com/ to make the ASCII art -- I haven't found a free alternative yet

**Comment by 2010YOUY01** [comment]
Previous review comments should be all addressed in recent commits, it's ready for review now.
Also filed an issue for the remaining tasks related to this PR.

**Review by alamb — APPROVED** [review]

**Comment by alamb** [comment]
Thanks @2010YOUY01  -- I took the liberty of merging up from main to resolve a conflict on this branch and I plan to merge it when the CI has passed

**Comment by alamb** [comment]
Thanks again @2010YOUY01

File: datafusion/core/src/datasource/physical_plan/csv.rs (hunk 7 above)
Comment (marvinlanhenke) [reply datafusion/core/src/datasource/physical_plan/csv.rs:403]: > # Alternate idea
> Another approach that would reduce the number of object store requests would be to read past the initial range and stop at the next newline `\n` like this:
> 
> ```
>   Each partition reads *more* than its assigned ranged to find                     
>   the trailing new line, and ignores everything afterwards                         
>                                                                                    
>                                                                                    
> ┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─                                                          
> ┌─────────────┬──┬───────┼────────────┬──┬────────────────┬──┬────────┬──┬────────┐
> │     ...     │\n│        ...         │\n│      ...       │\n│  ...   │\n│  ...   │
> └─────────────┴──┴───────┼────────────┴──┴────────────────┴──┴────────┴──┴────────┘
> │                                                                                  
>  ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘                                                         
>   Partition 0                                                                      
>   Read                                                                             
> ```
> 
> I think the tricky bit of this design would be to ensure enough extra data was read. Initially, maybe we could just pick something sufficiently large for most files, like 1MB and error if the next newline can't be found. As a follow on we could add some fanciness like make another object store request if necessary.

@alamb 
...working on #8502; I noticed the same thing - so am I right; that this "issue" is yet to be resolved? 

One could benchmark the different approaches (this PR vs. alternative Idea); I like the idea of picking a "sufficient large enough range at the beginning" and trying to recover with an additional object_store request.

Comment (alamb) [reply datafusion/core/src/datasource/physical_plan/csv.rs:403]: > I benchmarked it against 60mil rows of NDJSON - with no difference compared to the other approach. I could not gain any performance. Maybe my implementation is to naive? Or the overhead for the object store requests is to small when working with the local Filesystem?

Yes, I think this is the fundamental observation -- multiple requests to an actual remote object store is quite costly (like 10s of ms minimum) and fetching anything less than several MB in one request is likely less efficient than a single large request.

There is more about object storage in  [Exploiting Cloud Object Storage for High-Performance Analytics](https://www.vldb.org/pvldb/vol16/p2769-durner.pdf) : VLDB 2023 paper about running high performance analytics in object stores.

Comment (marvinlanhenke) [reply datafusion/core/src/datasource/physical_plan/csv.rs:403]: @alamb
Thanks for linking the paper.

So I'll guess the alternate approach is still worth pursuing?

I would proceed by testing the POC against a remote store and if this looks promising - I'd create a separate issue to discuss and refine the approach further?

Comment (alamb) [reply datafusion/core/src/datasource/physical_plan/csv.rs:403]: Sounds like a good plan to me

**Pull request merged**